src/base/YamlWriter.cpp

```diff
@@ -0,0 +1,217 @@
+    size_t nspecies_total = 0;
+    for (size_t i = 0; i < m_phases.size(); i++) {
+        m_phases[i]->thermo()->getParameters(phaseDefs[i]);
+        nspecies_total += m_phases[i]->thermo()->nSpecies();
+        const auto& kin = m_phases[i]->kinetics();
+        if (kin) {
+            kin->getParameters(phaseDefs[i]);
+            if (phaseDefs[i].hasKey("kinetics") && kin->nReactions() == 0) {
+                phaseDefs[i]["reactions"] = "none";
+            }
+        }
+        const auto& tran = m_phases[i]->transport();
+        if (tran) {
+            tran->getParameters(phaseDefs[i]);
+        }
+
+        if (!m_skip_user_defined) {
+            for (const auto& item : m_phases[i]->thermo()->input()) {
+                if (!phaseDefs[i].hasKey(item.first)) {
+                    phaseDefs[i][item.first] = item.second;
+                }
+            }
+        }
+    }
+    output["phases"] = phaseDefs;
+
+    // Build species definitions for all phases
+    std::vector<AnyMap> speciesDefs;
+    speciesDefs.reserve(nspecies_total);
+    std::unordered_map<std::string, size_t> speciesDefIndex;
+    for (const auto& phase : m_phases) {
+        const auto thermo = phase->thermo();
+        for (const auto& name : thermo->speciesNames()) {
+            const auto& species = thermo->species(name);
+            AnyMap speciesDef;
+            species->getParameters(speciesDef, !m_skip_user_defined);
+
+            thermo->getSpeciesParameters(name, speciesDef);
+            if (!m_skip_user_defined
+                && species->input.hasKey("equation-of-state")) {
+                auto& eosIn = species->input["equation-of-state"].asVector<AnyMap>();
+                for (const auto& eos : eosIn) {
+                    auto& out = speciesDef["equation-of-state"].getMapWhere(
+                        "model", eos["model"].asString(), true);
+                    for (const auto& item : eos) {
+                        if (!out.hasKey(item.first)) {
+                            out[item.first] = item.second;
+                        }
+                    }
+                }
+            }
+
+            if (!m_skip_user_defined) {
+                for (const auto& item : species->input) {
+                    if (!speciesDef.hasKey(item.first)) {
+                        speciesDef[item.first] = item.second;
+                    }
+                }
+            }
+
+            if (speciesDefIndex.count(name) == 0) {
+                speciesDefs.emplace_back(speciesDef);
+                speciesDefIndex[name] = speciesDefs.size() - 1;
+            } else if (speciesDefs[speciesDefIndex[name]] != speciesDef) {
+                throw CanteraError("YamlWriter::toYamlString",
+                    "Multiple species with different definitions are not "
+                    "supported:\n>>>>>>\n{}\n======\n{}\n<<<<<<\n",
+                    speciesDef.toYamlString(),
+                    speciesDefs[speciesDefIndex[name]].toYamlString());
```

Comment:
As `YamlWriter` is central to serialization, it may make sense to add a unit test for special cases  (here and elsewhere as marked by CodeCov).